### ISSUE TITLE
XGBoost cannot train properly in PyBroker's train_fn (only 25 iterations, but works fine outside with 575 iterations)

### ISSUE THREAD

**Issue by fhilo** [edtechre/pybroker#221]
Hello, I have an issue with training XGBoost in a custom train_fn. I am new to programming, so please forgive me if my expression is unclear.

### Environment
- pybroker: 1.2.12
- xgboost: 3.2.0
- Python: 3.13

### The Problem
I am using a custom `train_fn` for XGBoost with early stopping.
I split `train_data` into training and validation sets manually (time-based split, no shuffle).

**The problem is:**
- When I run the same XGBoost code **outside PyBroker**, it runs **575 iterations** with early stopping.
- When I run it **inside PyBroker’s train_fn**, it **only stops at 25 iterations** and does not train fully.
Model performance is very poor.

### Official Example
I followed the official example, which returns `model, [features]`:
```python
def train_slr(symbol, train_data, test_data):
    ...
    return model, ['cmma_20']
```

### My XGBoost train_fn
```python
def train_xgb(symbol, train_data, test_data):
    
    split_idx = int(len(train_data) * 0.8)
    train_split = train_data.iloc[:split_idx]
    val_split = train_data.iloc[split_idx:]

    features = ['mabool', 'NATR', ...]
    
    X_train = train_split[features]
    y_train = train_split['5d_label']
    X_val = val_split[features]
    y_val = val_split['5d_label']
    
    dtrain = xgb.DMatrix(X_train, label=y_train)
    dval = xgb.DMatrix(X_val, label=y_val)
    
    params = {
        "objective": "binary:logistic",
        "eval_metric": "auc",
        "max_depth": 4,
        "learning_rate": 0.0169,
        "subsample": 0.9480,
        "colsample_bytree": 0.9492,
        "reg_alpha": 0.4324,
        "reg_lambda": 0.2857,
        "random_state": 42
    }

    model = xgb.train(
        params=params,
        dtrain=dtrain,
        num_boost_round=10000,
        early_stopping_rounds=100,
        evals=[(dval, "val")],
        verbose_eval=False
    )

    return model, features
```

### My Questions
1. Why does XGBoost only run 25 iterations inside train_fn but 575 outside?
2. Is there any restriction or hidden behavior in train_fn that stops XGBoost early?
3. What is the correct way to use XGBoost with early stopping and validation set in PyBroker?
4. Is there something wrong with my train_fn?
5. Could you provide a working example of XGBoost train_fn with validation and early stopping?

Thank you very much!

---

**Comment by atharvajoshi01** [comment]
I think I know what's going on here. PyBroker's train_fn likely captures or redirects stdout/stderr, which interferes with XGBoost's early stopping callback. The early_stopping_rounds callback relies on the evaluation metric being printed and tracked internally, but if PyBroker wraps the training in a way that resets or interferes with the callback state, XGBoost's internal best_iteration tracker gets confused.

Try passing verbose_eval=False explicitly and check if the model object's best_iteration attribute is correct after training. If the iteration count is still wrong, the issue is probably in how PyBroker pickles/unpickles the model between calls.

**Comment by fhilo** [comment]
> I think I know what's going on here. PyBroker's train_fn likely captures or redirects stdout/stderr, which interferes with XGBoost's early stopping callback. The early_stopping_rounds callback relies on the evaluation metric being printed and tracked internally, but if PyBroker wraps the training in a way that resets or interferes with the callback state, XGBoost's internal best_iteration tracker gets confused.
> 
> Try passing verbose_eval=False explicitly and check if the model object's best_iteration attribute is correct after training. If the iteration count is still wrong, the issue is probably in how PyBroker pickles/unpickles the model between calls.

Thanks for your kind and helpful reply! I think I've found the problem. I had the dataset shuffled by default when splitting it, but shuffling is not allowed in quantitative finance. I should set `shuffle=False` instead. Until I gain a better understanding of these concepts, I think this issue can be closed. Thank you again for your warm help.

**Comment by edtechre** [comment]
Thank you @atharvajoshi01 !
